exporters/prometheus/exporter.go

```diff
@@ -140,48 +146,60 @@ func (c *collector) Collect(ch chan<- prometheus.Metric) {
 		var keys, values [2]string
 
 		if !c.disableScopeInfo {
-			scopeInfo, ok := c.scopeInfos[scopeMetrics.Scope]
-			if !ok {
-				scopeInfo, err = createScopeInfoMetric(scopeMetrics.Scope)
-				if err != nil {
-					otel.Handle(err)
+			func() {
+				c.mu.Lock()
+				defer c.mu.Unlock()
+
+				scopeInfo, ok := c.scopeInfos[scopeMetrics.Scope]
+				if !ok {
+					scopeInfo, err = createScopeInfoMetric(scopeMetrics.Scope)
+					if err != nil {
+						otel.Handle(err)
+					}
+					c.scopeInfos[scopeMetrics.Scope] = scopeInfo
 				}
-				c.scopeInfos[scopeMetrics.Scope] = scopeInfo
-			}
-			ch <- scopeInfo
+				ch <- scopeInfo
+			}()
```

Comment:
Sending to a channel inside a critical section looks akward. Is it really needed or can we move `ch <- scopeInfo` outside the critical section? E.g.

```golang
			var prometheus.Metric scopeInfo
			func() {
				c.mu.Lock()
				defer c.mu.Unlock()

				var ( // Declare variables to avoid shadowing.
					ok bool
					err error
				)
				if scopeInfo, ok = c.scopeInfos[scopeMetrics.Scope]; !ok {
					if scopeInfo, err = createScopeInfoMetric(scopeMetrics.Scope); err != nil
						otel.Handle(err)
					}
					c.scopeInfos[scopeMetrics.Scope] = scopeInfo
				}
			}()
			ch <- scopeInfo
```

Reply:
But probably we should change it to (based on my other comment)

```golang
			var ( // Declare variables to avoid shadowing.
				prometheus.Metric scopeInfo
				ok bool
				err error
			)

			c.mu.Lock()
			if scopeInfo, ok = c.scopeInfos[scopeMetrics.Scope]; !ok {
				scopeInfo, err = createScopeInfoMetric(scopeMetrics.Scope);
				c.scopeInfos[scopeMetrics.Scope] = scopeInfo
			}
			c.mu.Unlock()

			if err != nil {
				otel.Handle(err)
			}
			ch <- scopeInfo
```

Reply:
Maybe it could be further refactored so that the critical section is extracted to a separate method.

Reply:
> Maybe it could be further refactored so that the critical section is extracted to a separate method.

I like this one better. Will do.

Reply:
@pellared refactored, please take a look.